### PR TITLE
Test for card browser changing color when card flag changes

### LINKED ISSUE #8744: Add regression test for card browser changing color when card flag changes
This is follow on work for issue #8485

It was fixed in #8735 but without a test confirming the fix and making sure it doesn't happen again

This issue tracks creation of a test that verifies the issue before the fix from #8735 and makes sure it is actually fixed

In [a comment on the related PR](https://github.com/ankidroid/Anki-Android/pull/8735#pullrequestreview-649336285):

> Okay, is it worth obtaining logs from @ShridharGoel (or adding more logs if they're insufficient) so we can better understand this in the future?
> 
> Dealing with an unknown regression worries me: there's no reason that we won't reintroduce it at a later date
> 
> We shouldn't close the issue until it's verified to be fixed by @ShridharGoel , but this looks reasonable to me

### PR DESCRIPTION
**Pull request by dorrin-sot**
## Fixes
Fixes #8744

## Approach
After testing different approaches to reproduce the bug, I found that using a deck of about 75 cards minimum would cause the error.
After that I added the method `flagsAreShownInBigDecksTest` in `CardBrowserTest` class in which I selected and flagged a random card, unflagged the card, and in the end selected and flagged all the cards, testing each time that the flags are correct.

## How Has This Been Tested?
Checked out commit 3e798367 and verified that the test failed, then went back to normal branch, verifying the test pass.

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java [hunk 1]
```diff
@@ -260,6 +262,44 @@ public void changeDeckViaTaskIsHandledCorrectly() {
         }
     }
 
+
+    @Test
+    public void flagsAreShownInBigDecksTest() {
+        CardBrowser cardBrowser = getBrowserWithNotes(75);
+
+        // select a random card
+        cardBrowser.checkCardsAtPositions((int) (random() * 74));
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:271]: Either: 

* use a `Random` with a fixed seed
* Check 'em all
* record the seed

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java [hunk 2]
```diff
@@ -260,6 +262,44 @@ public void changeDeckViaTaskIsHandledCorrectly() {
...
+        // flag the selected card with flag = 1
+        cardBrowser.flagTask(flag);
+        shadowOf(getMainLooper()).idle();
+        // check if card flag turned to flag = 1
+        assertThat("Card should be flagged", getCheckedCard(cardBrowser).getCard().userFlag(), is(flag));
+
+        // unflag the selected card with flag = 0
+        final int unflagFlag = 0;
+        cardBrowser.flagTask(unflagFlag);
+        shadowOf(getMainLooper()).idle();
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:282]: Use: `RobolectricTest.advanceRobolectricLooper()`

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
minor change, looks great!

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java (hunk 1 above)
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:271]: Actually... note 3: class `Random` should have a `nextInt` which reads better than a multiplication

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java [hunk 3]
```diff
@@ -260,6 +262,44 @@ public void changeDeckViaTaskIsHandledCorrectly() {
         }
     }
 
+
+    @Test
+    public void flagsAreShownInBigDecksTest() {
+        CardBrowser cardBrowser = getBrowserWithNotes(75);
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:268]: extract the int, and use `numberOfNotes - 1` in the code below

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Thanks so much for the PR - a few more changes

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java (hunk 1 above)
Comment (dorrin-sot) [reply AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java]: I set the seed to 1 and used `assumeThat` to make sure that position gets checked

**Comment by david-allison-1** [comment]
Needs rebasing to remove irrelevant changes

**Comment by dorrin-sot** [comment]
> Needs rebasing to remove irrelevant changes

Done

**Comment by dorrin-sot** [comment]
Accidentally removed requested changes while rebasing now actually done

**Review by david-allison-1 — APPROVED** [review]
More testing! Lovely jubbly. Thanks!!!

**Review by mikehardy — APPROVED** [review]
Thank you!

**Comment by mikehardy** [comment]
Hi there!

Apologies for the delay, and if you're already submitted something to OpenCollective for PRs merged in June 2021 PRs merged, you may feel free to ignore this and it will be processed shortly

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for June 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged in May)

Thanks!

**Pull request merged**